core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java

```diff
@@ -3212,28 +3213,27 @@ public static ColumnarFormatVersion getFormatVersion(CarbonTable carbonTable)
       }
       storePath = carbonTable.getSegmentPath(validSegments.get(0).getSegmentNo());
     }
-
-    CarbonFile[] carbonFiles = FileFactory
-        .getCarbonFile(storePath)
-        .listFiles(new CarbonFileFilter() {
-          @Override
-          public boolean accept(CarbonFile file) {
-            if (file == null) {
-              return false;
-            }
-            return file.getName().endsWith("carbondata");
-          }
-        });
-    if (carbonFiles == null || carbonFiles.length < 1) {
-      return CarbonProperties.getInstance().getFormatVersion();
+    // get the carbon index file header
+    FileFactory.FileType fileType = FileFactory.getFileType(storePath);
+    ColumnarFormatVersion version = null;
+    if (FileFactory.isFileExist(storePath, fileType)) {
+      SegmentIndexFileStore fileStore = new SegmentIndexFileStore();
+      fileStore.readAllIIndexOfSegment(storePath);
+      Map<String, byte[]> carbonIndexMap = fileStore.getCarbonIndexMap();
+      if (carbonIndexMap.size() == 0) {
+        version = CarbonProperties.getInstance().getFormatVersion();
+      }
+      CarbonIndexFileReader indexReader = new CarbonIndexFileReader();
+      for (byte[] fileData : carbonIndexMap.values()) {
+        indexReader.openThriftReader(fileData);
+        IndexHeader indexHeader = indexReader.readIndexHeader();
+        version = ColumnarFormatVersion.valueOf((short)indexHeader.getVersion());
+        break;
+      }
+    } else {
+      version = CarbonProperties.getInstance().getFormatVersion();
```

Comment:
1. This else condition is not required as for a valid segment its path will exist in the file system.
2. If at all the path does not exist then read the version info from the next valid segment and log a warning that valid segment path does not exist in the system

Reply:
modified according to review comments